effect/shared/src/main/scala/scalaz/effect/IO.scala

```diff
@@ -0,0 +1,549 @@
+   *
+   * {{{
+   * openFile("data.json").bracket(closeFile(_)) { file =>
+   *   for {
+   *     header <- readHeader(file)
+   *     ...
+   *   } yield result
+   * }
+   * }}}
+   */
+  final def bracket[B](release: A => IO[Unit])(use: A => IO[B]): IO[B] = IO.Bracket(this, (_: BracketResult[B], a: A) => release(a), use)
+
+  /**
+   * A more powerful version of `bracket` that provides information on whether
+   * or not `use` succeeded to the release action.
+   */
+  final def bracket0[B](release: (BracketResult[B], A) => IO[Unit])(use: A => IO[B]): IO[B] = IO.Bracket(this, release, use)
+
+  /**
+   * A less powerful variant of `bracket` where the value produced by this
+   * action is not needed.
+   */
+  final def bracket_[B](release: IO[Unit])(use: IO[B]): IO[B] =
+    bracket(_ => release)(_ => use)
+
+  /**
+   * Executes the specified finalizer, whether this action succeeds, fails, or
+   * is interrupted.
+   */
+  final def ensuring(finalizer: IO[Unit]): IO[A] =
+    IO.unit.bracket(_ => finalizer)(_ => self)
+
+  /**
+   * Executes the release action only if there was an error.
+   */
+  final def bracketOnError[B](release: A => IO[Unit])(use: A => IO[B]): IO[B] =
+    bracket0((r: BracketResult[B], a: A) => r match {
+      case BracketResult.Failed(_) => release(a)
+      case BracketResult.Interrupted(_) => release(a)
+      case _ => IO.unit
+    })(use)
+
+  /**
+   * Runs the specified cleanup action if this action errors, providing the
+   * error to the cleanup action. The cleanup action will not be interrupted.
+   */
+  final def onError(cleanup: Throwable => IO[Unit]): IO[A] = IO.unit.bracket0((r: BracketResult[A], a: Unit) => r match {
+    case BracketResult.Failed(e) => cleanup(e)
+    case BracketResult.Interrupted(e) => cleanup(e)
+    case _ => IO.unit
+  })(_ => self)
+
+  /**
+   * Supervises this action, which ensures that any fibers that are forked
+   * by the action are killed with the specified error when this action
+   * completes.
+   */
+  final def supervised(error: Throwable): IO[A] = IO.Supervise(self, error)
+
+  /**
+   * Performs this action non-interruptibly. This will prevent the
+   * action from being killed externally, but the action may fail
+   * for internal reasons (e.g. an uncaught exception).
+   */
+  final def uninterruptibly: IO[A] = IO.Uninterruptible(self)
+
+  /**
+   * Recovers from all errors.
+   *
+   * {{{
+   * openFile("config.json").catchAll(_ => IO.now(defaultConfig))
+   * }}}
+   */
+  final def catchAll(h: Throwable => IO[A]): IO[A] = catchSome {
+    case t : Throwable => h(t)
+  }
+
+  /**
+   * Recovers from some or all of the error cases.
+   *
+   * {{{
+   * openFile("data.json").catchSome {
+   *   case FileNotFoundException(_) => openFile("backup.json")
+   * }
+   * }}}
+   */
+  final def catchSome(pf: PartialFunction[Throwable, IO[A]]): IO[A] = {
+    def tryRescue(t: Throwable): IO[A] = t match {
+      case t if pf.isDefinedAt(t) => pf(t)
```

Comment:
could be a simple `if`/`else` instead of a guard.

Reply:
Fixed. 👍